### PR TITLE
PMM-7292 scrape from server pmm agent

### PR DESCRIPTION
**Pull request by rishat-ishbulatov**
PMM-7292

Build: SUBMODULES-0

- [ ] Links to other linked pull requests (optional).

### CODE REVIEW

File: managed/services/victoriametrics/victoriametrics.go [hunk 1]
```diff
@@ -416,3 +416,21 @@ func (svc *Service) IsReady(ctx context.Context) error {
...
+		ServiceDiscoveryConfig: config.ServiceDiscoveryConfig{
+			StaticConfigs: []*config.Group{
+				{
+					Targets: []string{"127.0.0.1:7777"},
+					Labels:  map[string]string{"instance": models.PMMServerAgentID},
+				},
+			},
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply managed/services/victoriametrics/victoriametrics.go:436]: I don't see the difference between this and the one in scrape_config.go

File: managed/services/victoriametrics/scrape_configs.go [hunk 2]
```diff
@@ -590,5 +590,23 @@ func scrapeConfigsForVMAgent(s *models.MetricsResolutions, params *scrapeConfigP
 			Labels:  labels,
 		}},
 	}
-	return []*config.ScrapeConfig{cfg}, nil
+	return []*config.ScrapeConfig{cfg, scrapeConfigForPmmAgent(s, params)}, nil
```
Comment (BupycHuk) [inline managed/services/victoriametrics/scrape_configs.go:593]: probably `scrapeConfigForPmmAgent` shouldn't be called from here. let's call it from the place where `scrapeConfigsForVMAgent` is called.

**Review by BupycHuk — APPROVED** [review]

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Sorry I approved by mistake

File: managed/services/victoriametrics/prometheus.go [hunk 3]
```diff
@@ -209,9 +217,68 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+	}
+
+	// find Service for this Agent
+	var paramsService *models.Service
+	if agent.ServiceID != nil {
+		paramsService, err = models.FindServiceByID(q, pointer.GetString(agent.ServiceID))
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (BupycHuk) [inline managed/services/victoriametrics/prometheus.go:247]: serviceID and NodeID are always nil for pmm-agent, so we can remove everything related to ServiceID and NodeID. it also means that paramsNode is always nil.

File: managed/services/victoriametrics/prometheus.go [hunk 4]
```diff
@@ -209,9 +217,68 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+		paramsNode, err = models.FindNodeByID(q, paramsService.NodeID)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	pmmAgentNode := &models.Node{NodeID: pointer.GetString(agent.RunsOnNodeID)}
+	if err = q.Reload(pmmAgentNode); err != nil {
+		return nil, errors.WithStack(err)
+	}
```
Comment (BupycHuk) [inline managed/services/victoriametrics/prometheus.go:264]: let's use `models.FindNodeByID`

File: managed/services/victoriametrics/prometheus.go [hunk 5]
```diff
@@ -209,9 +217,68 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+
+	paramsHost := pmmAgentNode.Address
+	paramPMMAgentVersion, err := version.Parse(pointer.GetString(agent.Version))
+	if err != nil {
+		l.Warnf("couldn't parse pmm-agent version for pmm-agent %s: %q", agent.AgentID, err)
+	}
+
+	scfg, err := scrapeConfigForPmmAgent(s, &scrapeConfigParams{
+		host:            paramsHost,
+		node:            paramsNode,
```
Comment (BupycHuk) [inline managed/services/victoriametrics/prometheus.go:274]: let's use `pmmAgentNode` here

File: managed/services/victoriametrics/prometheus.go [hunk 6]
```diff
@@ -209,9 +217,68 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+		service:         paramsService,
```
Comment (BupycHuk) [inline managed/services/victoriametrics/prometheus.go:275]: paramsService is nil

**Review by BupycHuk — COMMENTED** [review]
great, let's cleanup a little bit.

File: managed/services/victoriametrics/prometheus.go [hunk 7]
```diff
@@ -209,9 +217,42 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+
+	// find Node for this Agent
+	node, err := models.FindNodeByID(q, pointer.GetString(agent.RunsOnNodeID))
+	if err != nil {
+		return nil, err
+	}
+
+	pmmAgentVersion, err := version.Parse(pointer.GetString(agent.Version))
+	if err != nil {
+		l.Warnf("couldn't parse pmm-agent version for pmm-agent %s: %q", agent.AgentID, err)
```
Comment (artemgavrilov) [inline managed/services/victoriametrics/prometheus.go:243]: I guess we need return here

File: managed/services/victoriametrics/prometheus.go [hunk 8]
```diff
@@ -209,9 +217,42 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+
+	scfg, err := scrapeConfigForPmmAgent(s, &scrapeConfigParams{
+		host:            node.Address,
+		node:            node,
+		service:         nil,
+		agent:           agent,
+		pmmAgentVersion: pmmAgentVersion,
+	})
+	if err != nil {
+		l.Warnf("Failed to add %s %q, skipping: %s.", agent.AgentType, agent.AgentID, err)
```
Comment (artemgavrilov) [inline managed/services/victoriametrics/prometheus.go:254]: And here

File: managed/services/victoriametrics/prometheus.go (hunk 7 above)
Comment (artemgavrilov) [reply managed/services/victoriametrics/prometheus.go:243]: Hm, no, probably here we can continue without return. But do we use `pmmAgentVersion` somewhere at all?

**Review by artemgavrilov — APPROVED** [review]

File: managed/services/victoriametrics/prometheus.go [hunk 9]
```diff
@@ -209,9 +213,55 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
...
+		scrapeConfigForQANAPI2(s.MR),
+		scrapeConfigForServerPMMAgent(s.MR))
 	// TODO Refactor to remove boolean positional parameter when Prometheus is removed
 	if dbaas {
 		cfg.ScrapeConfigs = append(cfg.ScrapeConfigs, scrapeConfigForDBaaSController(s.MR))
 	}
 }
+
+// AddPmmAgentScrapeConfig adds pmm-agent metrics to scrape targets.
+func AddPmmAgentScrapeConfig(cfg *config.Config, l *logrus.Entry, q *reform.Querier, s *models.MetricsResolutions, pmmAgentID *string) error {
```
Comment (BupycHuk) [inline managed/services/victoriametrics/prometheus.go:225]: let's make it private since we don't use this method outside of this package.

Comment (BupycHuk) [inline managed/services/victoriametrics/prometheus.go:225]: can we return a scrape config here and add it in the method above it should make the code more consistent

**Review by BupycHuk — COMMENTED** [review]
Everything looks good, just a minor changes

File: managed/services/agents/registry.go [hunk 10]
```diff
@@ -197,12 +197,25 @@ func (r *Registry) register(stream agentpb.Agent_ConnectServer) (*pmmAgentInfo,
...
+		// port with default value or not changed - skipping
+		if agentMD.MetricsPort == 0 || (agent.ListenPort != nil && *agent.ListenPort == agentMD.MetricsPort) {
+			return nil
+		}
+
+		// registering new pmmAgent port in database for scrape config.
+		params := models.ChangeCommonAgentParams{Port: &agentMD.MetricsPort}
+		if _, err = models.ChangeAgent(tx.Querier, agentMD.ID, &params); err != nil {
+			return err
+		}
```
Comment (BupycHuk) [inline managed/services/agents/registry.go:218]: let's make this update inside `authenticate` function somewhere here https://github.com/percona/pmm/pull/1237/files#diff-61640292d442a75c3b3b3aa4944a99aef9860ea2325e8c9bc86cbb00a627f6f0R292

Comment (BupycHuk) [reply managed/services/agents/registry.go:218]: in that case you won't need to return `agent` and rename `agent` to `agentInfo` below.

**Review by BupycHuk — COMMENTED** [review]
Looks good to me, just one improvement.

File: managed/services/agents/registry.go (hunk 10 above)
Comment (rishat-ishbulatov) [reply managed/services/agents/registry.go]: ok

Comment (rishat-ishbulatov) [reply managed/services/agents/registry.go]: done

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**